src/datasets/arrow_dataset.py

```diff
@@ -2818,6 +2760,40 @@ def _map_single(
         # If set to False, no new arrow table will be created
         update_data = None
 
+        lazy_formatting = not input_columns and supports_lazy_formatting(self._format_type)
+        input_formatter = get_formatter(
+            self._format_type, features=self.features, lazy=lazy_formatting, **self._format_kwargs
+        )
+        if lazy_formatting:
+            # Save undecoded data (to preserve exact image/audio format for instance) after processing unless decoded data is modified in the `function`
+            class DictWithDecodableDataToRestore(LazyPythonRow if not batched else LazyPythonBatch):
+                def _init(self):
+                    self.decodable_data_to_restore = {}
+
+                def __getitem__(self, key):
+                    if (
+                        key in self.keys_to_format
+                        and self.formatter.features._column_requires_decoding[key]
+                        and key not in self.decodable_data_to_restore
+                    ):
+                        self.decodable_data_to_restore[key] = self.data[key]
+                    return super().__getitem__(key)
+
+                def __setitem__(self, key, value):
+                    if key in self.decodable_data_to_restore:
+                        del self.decodable_data_to_restore[key]
+                    super().__setitem__(key, value)
+
+                def __delitem__(self, key):
+                    if key in self.decodable_data_to_restore:
+                        del self.decodable_data_to_restore[key]
+                    super().__delitem__(key)
```

Comment:
(nit) Maybe this can be integrated in LazyDict Directly ? This way we don't need an extra class.

Consider this example

```python
t = pa.table({"image": ...})
batch = LazyBatch(t, formatter=NumpyFormatter())
batch["shape"] = [img.shape for img in batch["image"]]
writer.write_table(batch.to_arrow())
```

Under the hood `batch.to_arrow()` would reuse the arrow data from `t` for "image", and make new arrow data for "shape". This way we avoid one conversion from python to arrow - the same way DictWithDecodableDataToRestore works right now.

Let me know if it makes sense

Reply:
This API has been removed, as we no longer need it (and it's somewhat hard to understand)